tests/unit/recommenders/models/test_sar_singlenode.py

```diff
@@ -413,3 +418,41 @@ def test_match_similarity_type_from_json_file(header):
 
     # make sure fit still works when similarity type is loaded from a json file
     model.fit(train)
+
+def test_get_user_and_item_based_topk(header):
+    model = SARSingleNode(**header)
+    train = pd.DataFrame(
+        {
+            header["col_user"]: [1, 1, 2, 2, 3, 3, 3, 3, 4, 4],
+            header["col_item"]: [1, 2, 1, 3, 3, 4, 5, 6, 1, 2],
+            header["col_rating"]: [3.0, 4.0, 5.0, 4.0, 3.0, 2.0, 1.0, 5.0, 1.0, 1.0]
+        }
+    )
+    model.fit(train, compute_user_similarity=True)
+
+    similar_users = model.get_user_based_topk(pd.DataFrame({header['col_user']: [1]}))
+    assert similar_users[header['col_user']].iloc[0] == 4
+
+    similar_users = model.get_user_based_topk(pd.DataFrame({header['col_user']: [1, 2]}))
+    assert similar_users[header['col_user']].iloc[0] == 4
```

Comment:
I've been doing some debugging, if I do:
```
train = pd.DataFrame(
        {
            header["col_user"]: [1, 1, 2, 2, 3, 3, 3, 3, 4, 4],
            header["col_item"]: [1, 2, 1, 3, 3, 4, 5, 6, 1, 2],
            header["col_rating"]: [3.0, 4.0, 5.0, 4.0, 3.0, 2.0, 1.0, 5.0, 1.0, 1.0]
        }
    )
```
I get:
```
	UserId	MovieId	prediction
0	4	0	1.000000
1	2	0	0.333333
2	3	0	0.000000
```
but if I do the following, making the user 1 and 2 the same:
```
train = pd.DataFrame(
        {
            header["col_user"]: [1, 1, 2, 2, 3, 3, 3, 3, 4, 4],
            #header["col_item"]: [1, 2, 1, 3, 3, 4, 5, 6, 1, 2],
            #header["col_rating"]: [3.0, 4.0, 5.0, 4.0, 3.0, 2.0, 1.0, 5.0, 1.0, 1.0]
            header["col_item"]: [1, 2, 1, 2, 3, 4, 5, 6, 1, 2],
            header["col_rating"]: [3.0, 4.0, 3.0, 4.0, 3.0, 2.0, 1.0, 5.0, 5.0, 1.0]
        }
    )
```
I get:
UserId | MovieId | prediction
-- | -- | --
2 | 0 | 1.0
4 | 0 | 1.0
3 | 0 | 0.0

But user 2 is equal to user 1, and it gets a prediction of 1, which is the same as user 4, that has seen the same movies as user 1, but rated them completely different. User 2 should get a higher score than user 4 in this case. Do you know why this is happening?

Reply:
I believe this is because user similarity is calculated from user cooccurrence, which does not actually look at the ratings.  The same is true for items, so I believe you would see the same issue with items.